# Market Basket Analysis

In [1]:
pip install mlxtend

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.3.1 -> 23.2.1
[notice] To update, run: C:\Users\kumar\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [2]:
#Exporting required Libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import gc
import os
import time
import datetime as dt
import mlxtend.preprocessing
import mlxtend.frequent_patterns

In [3]:
#Reading the file
df=pd.read_csv('Merged_data_file.csv')

### Note-Performing the market basket analysis only for 2023 data for the period of 6 months as the process time of entire data takes long time

In [10]:
#MBA_df stands for market basket aalysis data frame
#narrowing down the data to the required number of columns
df_2023=df[df['year']==2023]

MBA_df=df_2023[['Order_Number','Lineitem_name']]

In [11]:
# the data is already cleaned in the provious EDA notebook

In [12]:
# converting the basket item to list of lists
item_list = []

for num in list(set(MBA_df.Order_Number.tolist())):
    # filter data set down to one invoice number
    tmp_df = MBA_df.loc[MBA_df['Order_Number'] == num]
    # extract item descriptions and convert to list
    tmp_items = tmp_df.Lineitem_name.tolist()
    # append list invoice_item_list
    item_list.append(tmp_items)

print(item_list[1:3])

[['Sunfeast All Rounder Biscuits : Thin Potato Biscuits - Cream & Herb, 28.2g', 'Sunfeast Caker Trinity Cake, Triple Chocolate, 25g', 'B Natural Select - Tender Coconut Water, 200ml X 6 bottles', 'Aashirvaad Svasti Pure Cow Ghee 1L', 'Sunfeast Farmlite 5 Seed Digestive Biscuit 100g, High fibre, Goodness of 5 Power Seeds and Wheat Fibre', 'Sunfeast Farmlite Nuts Digestive Biscuit 100g, High fibre, Goodness of Almonds, Cashews and wheat fibre', 'Sunfeast Farmlite Active Oats with Almonds Biscuits 150g', 'Sunfeast Farmlite Active Oats with Raisins Biscuits 150g', 'Sunfeast Farmlite 5 Grain Digestive', 'Sunfeast Farmlite Active Oats with Chocolate Biscuits 150g', 'Aashirvaad Turmeric Powder 200g', 'Aashirvaad Vermicelli 400g', 'Aashirvaad Atta with Multigrains 5kg', 'Shower to Shower Sandal 150g', 'Nimeasy Dishwash Liquid Gel 250ml, Kitchen Utensil Cleaner, Lift Off Action with Enzyme Technology', 'Nimyle Herbal Floor Cleaner - 2 l', 'Mangaldeep Temple Diya Silver Tradition Agarbatti 50 St

In [13]:
# fitting the transaction encoder
online_encoder = mlxtend.preprocessing.TransactionEncoder()
online_encoder_array = online_encoder.fit_transform(item_list)

# converting the array as a dataframe
online_encoder_df = pd.DataFrame(online_encoder_array, columns=online_encoder.columns_)

# Print the results
online_encoder_df


AASHIRVAAD INSTANT MEALS SUJI HALWA WITH JAGGERY CUP 70G  \
0                                                  False          
1                                                  False          
2                                                  False          
3                                                  False          
4                                                  False          
...                                                  ...          
37360                                              False          
37361                                              False          
37362                                              False          
37363                                              False          
37364                                              False          

       AASHIRVAAD SVASTI MISHTI DOI 400g  \
0                                  False   
1                                  False   
2                                  False   
3                                  False   
4                                  False   
...                                  ...   
37360                              False   
37361                              False   
37362                              False   
37363                              False   
37364                              False   

       AASHIRVAAD SVASTI MISHTI DOI 85g (Pack of 6)  \
0                                             False   
1                                             False   
2                                             False   
3                                             False   
4                                             False   
...                                             ...   
37360                                         False   
37361                                         False   
37362                                         False   
37363                                         False   
37364                                         False   

       AASHIRVAAD SVASTI PANEER 190g  Aashirvaad  Malabar paratha, 400g  \
0                              False                              False   
1                              False                              False   
2                              False                               True   
3                              False                              False   
4                              False                              False   
...                              ...                                ...   
37360                          False                              False   
37361                          False                              False   
37362                          False                              False   
37363                          False                              False   
37364                          False                              False   

       Aashirvaad Aloo Paratha, 400g  Aashirvaad Atta with Multigrains 1kg  \
0                              False                                 False   
1                              False                                 False   
2                              False                                 False   
3                              False                                 False   
4                              False                                 False   
...                              ...                                   ...   
37360                          False                                 False   
37361                          False                                 False   
37362                          False                                 False   
37363                          False                                 False   
37364                          False                                 False   

       Aashirvaad Atta with Multigrains 5kg  Aashirvaad Bansi Rava, 1Kg  \
0                                     False                       False   
1                       

# Association Rules with Apriori Algorithm

#### Using Apriori algorithm to analyse the frequent itemsets in this transaction data

In [14]:
#running the algorithm with minimum support of 0.0.1 as default is by 0.5

apriori_model = mlxtend.frequent_patterns.apriori(online_encoder_df, min_support=0.01)
apriori_model

support         itemsets
0    0.032356              (4)
1    0.026495              (5)
2    0.012338              (6)
3    0.030001              (7)
4    0.011107              (9)
..        ...              ...
331  0.010571  (849, 873, 847)
332  0.011026  (903, 849, 847)
333  0.011481  (873, 871, 847)
334  0.010250  (872, 873, 847)
335  0.012418  (872, 873, 871)

[336 rows x 2 columns]

In [15]:
#Rinning the algorithm again with column names(item names) replacing the numerical designations
apriori_model_colnames = mlxtend.frequent_patterns.apriori(
    online_encoder_df, 
    min_support=0.01,
    use_colnames=True
)

apriori_model_colnames

support                                           itemsets
0    0.032356                (Aashirvaad  Malabar paratha, 400g)
1    0.026495                    (Aashirvaad Aloo Paratha, 400g)
2    0.012338             (Aashirvaad Atta with Multigrains 1kg)
3    0.030001             (Aashirvaad Atta with Multigrains 5kg)
4    0.011107                      (Aashirvaad Bansi Rava, 500g)
..        ...                                                ...
331  0.010571  (Sunfeast Caker Trinity Cake, Triple Chocolate...
332  0.011026  (Sunfeast Caker Trinity Cake, Triple Chocolate...
333  0.011481  (Sunfeast Farmlite Active Oats with Almonds Bi...
334  0.010250  (Sunfeast Farmlite Active Oats with Chocolate ...
335  0.012418  (Sunfeast Farmlite Active Oats with Chocolate ...

[336 rows x 2 columns]

In [16]:
#adding another column that tells us the size itemsets
apriori_model_colnames['length'] = (
    apriori_model_colnames['itemsets'].apply(lambda x: len(x))
)

apriori_model_colnames

support                                           itemsets  length
0    0.032356                (Aashirvaad  Malabar paratha, 400g)       1
1    0.026495                    (Aashirvaad Aloo Paratha, 400g)       1
2    0.012338             (Aashirvaad Atta with Multigrains 1kg)       1
3    0.030001             (Aashirvaad Atta with Multigrains 5kg)       1
4    0.011107                      (Aashirvaad Bansi Rava, 500g)       1
..        ...                                                ...     ...
331  0.010571  (Sunfeast Caker Trinity Cake, Triple Chocolate...       3
332  0.011026  (Sunfeast Caker Trinity Cake, Triple Chocolate...       3
333  0.011481  (Sunfeast Farmlite Active Oats with Almonds Bi...       3
334  0.010250  (Sunfeast Farmlite Active Oats with Chocolate ...       3
335  0.012418  (Sunfeast Farmlite Active Oats with Chocolate ...       3

[336 rows x 3 columns]

In [18]:
#generating a rule with key parameters. Here, we are using confidence as the metric with 0.06
rules = mlxtend.frequent_patterns.association_rules(
    apriori_model_colnames, 
    metric="confidence",
    min_threshold=0.06, 
    support_only=False
)

rules

antecedents  \
0                  (Aashirvaad  Malabar paratha, 400g)   
1             (Aashirvaad Garlic Coriander Naan, 400g)   
2                  (Aashirvaad  Malabar paratha, 400g)   
3                  (Aashirvaad  Malabar paratha, 400g)   
4               (Aashirvaad Atta with Multigrains 5kg)   
..                                                 ...   
206  (Sunfeast Farmlite Active Oats with Chocolate ...   
207  (Sunfeast Farmlite Active Oats with Almonds Bi...   
208  (Sunfeast Farmlite Active Oats with Chocolate ...   
209  (Sunfeast Farmlite Active Oats with Almonds Bi...   
210  (Sunfeast Farmlite Active Oats with Raisins Bi...   

                                           consequents  antecedent support  \
0             (Aashirvaad Garlic Coriander Naan, 400g)            0.032356   
1                  (Aashirvaad  Malabar paratha, 400g)            0.030751   
2     (Sunfeast Caker Swiss Roll Cake, Chocolate, 25g)            0.032356   
3    (Sunfeast Caker Trinity Cake, Triple Chocolate...            0.032356   
4     (Sunfeast Caker Swiss Roll Cake, Chocolate, 25g)            0.030001   
..                                                 ...                 ...   
206  (Sunfeast Farmlite Active Oats with Almonds Bi...            0.022856   
207  (Sunfeast Farmlite Active Oats with Chocolate ...            0.025746   
208  (Sunfeast Farmlite Active Oats with Almonds Bi...            0.043570   
209  (Sunfeast Farmlite Active Oats with Chocolate ...            0.054704   
210  (Sunfeast Farmlite Active Oats with Chocolate ...            0.050582   

     consequent support   support  confidence       lift  leverage  \
0              0.030751  0.010571    0.326716  10.624677  0.009576   
1              0.032356  0.010571    0.343777  10.624677  0.009576   
2              0.228422  0.011615    0.358974   1.571538  0.004224   
3              0.208377  0.010009    0.309347   1.484554  0.003267   
4              0.228422  0.010732    0.357716   1.566031  0.003879   
..                  ...       ...         ...        ...       ...   
206            0.054704  0.012418    0.543326   9.932171  0.011168   
207            0.043570  0.012418    0.482328  11.070150  0.011296   
208            0.025746  0.012418    0.285012  11.070150  0.011296   
209            0.022856  0.012418    0.227006   9.932171  0.011168   
210            0.017958  0.012418    0.245503  13.670948  0.011510   

     conviction  zhangs_metric  
0      1.439585       0.936171  
1      1.474566       0.934620  
2      1.203661       0.375842  
3      1.146195       0.337311  
4      1.201304       0.372622  
..          ...            ...  
206    2.069957       0.920352  
207    1.847561       0.933706  
208    1.362616       0.951107  
209    1.264103       0.951360  
210    1.301584       0.976232  

[211 rows x 10 columns]

In [20]:
#number of associations
print("Number of Associations: {}".format(rules.shape[0]))

Number of Associations: 211


In [27]:
#setting metrics to get the most frequent brought  itemset
rules[ (rules['support'] >= 0.02) &
       (rules['confidence'] > 0.06) &
       (rules['lift'] > 6.0) ]

antecedents  \
109   (Sunfeast Caker Swiss Roll Cake, Chocolate, 23g)   
110  (Sunfeast Caker Trinity Cake, Triple Chocolate...   
146  (Sunfeast Farmlite Active Oats with Almonds Bi...   
147  (Sunfeast Farmlite Active Oats with Raisins Bi...   
148  (Sunfeast Farmlite Active Oats with Chocolate ...   
149  (Sunfeast Farmlite Active Oats with Raisins Bi...   

                                           consequents  antecedent support  \
109  (Sunfeast Caker Trinity Cake, Triple Chocolate...            0.045283   
110   (Sunfeast Caker Swiss Roll Cake, Chocolate, 23g)            0.040278   
146  (Sunfeast Farmlite Active Oats with Raisins Bi...            0.054704   
147  (Sunfeast Farmlite Active Oats with Almonds Bi...            0.050582   
148  (Sunfeast Farmlite Active Oats with Raisins Bi...            0.043570   
149  (Sunfeast Farmlite Active Oats with Chocolate ...            0.050582   

     consequent support   support  confidence       lift  leverage  \
109            0.040278  0.023899    0.527778  13.103267  0.022075   
110            0.045283  0.023899    0.593355  13.103267  0.022075   
146            0.050582  0.025746    0.470646   9.304593  0.022979   
147            0.054704  0.025746    0.508995   9.304593  0.022979   
148            0.050582  0.022856    0.524570  10.370666  0.020652   
149            0.043570  0.022856    0.451852  10.370666  0.020652   

     conviction  zhangs_metric  
109    2.032352       0.967494  
110    2.347793       0.962449  
146    1.793540       0.944176  
147    1.925227       0.940077  
148    1.996967       0.944737  
149    1.744838       0.951714

### Interpretation 

#### Setting the minimum threshold or metric values for confidence , support and lift totally depends on the business objective and requirements.  I have chosen these values with my best knowledge.

#### With the support values greater than  0.02,confidence greater than  0.06 and lift greated than 6 we find "Sunfeast Caker swiss roll caker chocolate" and "Sunfeast caker trinity cake" to the most frequently brought itemset.

#### The above table gives us the most frequently brought item set with the respective metric value

### The above result explains the best combination products. Taking row 109 as an example, the support value of 0.04 means there is 4% of occurrence of the combination product of antecedent and consequent in the total number of transaction, which means out of 100 baskets,4 combinations are of product A and Product B. Confidence value of 0.52 means , given that antecedent is present in the basket there are 52% of chance that the consequent will be present in the basket. A lift value of 13 means , presence of antecedent has uplifted the chances of presence of consequence by 13 times.

